src/data/finset/basic.lean

```diff
@@ -1994,6 +1994,24 @@ theorem card_image_of_inj_on [decidable_eq β] {f : α → β} {s : finset α}
   (H : ∀x∈s, ∀y∈s, f x = f y → x = y) : card (image f s) = card s :=
 by simp only [card, image_val_of_inj_on H, card_map]
 
+theorem inj_on_of_card_image_eq [decidable_eq β] {f : α → β} {s : finset α}
+  (H : card (image f s) = card s) :
+∀x∈s, ∀y∈s, f x = f y → x = y :=
```

Comment:
```suggestion
  ∀x∈s, ∀y∈s, f x = f y → x = y :=
```

Reply:
Or better:
```suggestion
   set.inj_on f s 
```

Reply:
Ah, I see the problem; this is a `finset` so `set.inj_on` doesn't work.

Reply:
Yeah I originally wanted this too but unfortunately it's not this simple

Reply:
`set.inj_on f ↑s` works and it's defeq, but the existing pattern (eg `card_image_of_inj_on`) was to avoid this

Reply:
Yeah, maybe just do @bryangingechen's hidden whitespace comment then

Reply:
```suggestion
  ∀ ⦃x⦄, x∈s → ∀ ⦃y⦄, y∈s → f x = f y → x = y :=
```
to match the implicitness of `inj_on`.

Reply:
What about matching `card_image_of_inj_on` and all the other finset lemmas which use this?

Reply:
`card_image_of_inj_on` has the binders in a hypothesis so they're irrelevant; it's only binders in the lemma itself that have an effect.

Reply:
Indeed, `inj_on_of_surj_on_of_card_le` from later in this file does indeed use this type of binder

Reply:
I'm really confused what you want here - you want the binders to match `set.inj_on` but not use `set.inj_on` even though that works? And I'm pretty sure `⦃⦄` binders do have an effect in hypotheses, I've seen Lean behave differently depending on whether or not they're there - you can see `finset.induction_on` for a simple example of using these in  a hypothesis. And finally, are you suggesting making the same change in `card_image_eq_iff_inj_on` as well?

Reply:
> I'm really confused what you want here

Sorry about that!

> - you want the binders to match `set.inj_on`

Yes

> but not use `set.inj_on` even though that works?

Yes, because that puts `x ∈ ↑s` in the arguments instead of `x ∈ s` , which while defeq is likely to be a pain somewhere.
If we already do that elsewhere, then feel free to just do it anyway.

> And I'm pretty sure `⦃⦄` binders do have an effect in hypotheses, I've seen Lean behave differently depending on whether or not they're there - you can see `finset.induction_on` for a simple example of using these in  a hypothesis.

They definitely make a difference in hypotheses to the _proof_ that uses them, but I'd want to see an example to be convinced that they make a difference to the caller.
Having said that, I wouldn't object to you adjusting the binders elsewhere in the file; I just think that the ones in hypotheses won't really make a difference.

> And finally, are you suggesting making the same change in `card_image_eq_iff_inj_on` as well? 

Yes, good point.

Reply:
I think using `set.inj_on` is the cleanest here, so I've just changed to that.

Reply:
It doesn't look like you pushed that change?

Reply:
Oops, thanks